### PR TITLE
Remove unnecessary buffer allocations in VCL

### PR DESCRIPTION
**Pull request by dmartin13**
# Description

- Some buffer allocations in the begin() and getRegionIterator() functions could be removed since _particlesToAdd contains only particles in rebuild iterations. So in all non-rebuild iterations the additionalVectors can be passed directly to the ContainerIterator without concatenating them with _particlesToAdd.
- Description added why it is (at the moment) necessary to keep the _particlesToAdd buffer-structure for VCL.

## Resolved Issues

- [x] solves the problem of memory allocations mentioned in #734

# How Has This Been Tested?

- All existing tests

### CODE REVIEW

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 1]
```diff
@@ -383,57 +404,135 @@ class VerletClusterLists : public ParticleContainerInterface<Particle>, public i
...
+
+  /**
+   * Helper function for begin() and getRegionIterator() that creates a ContainerIterator from all particles in
+   * _particlesToAdd
+   *
+   * @tparam regionIter
+   * @param additionalVectorsToPass
+   */
+  template <bool regionIter>
+  void createIteratorFromParticlesToAdd(
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterLists/VerletClusterLists.h:438]: I feel this name is somewhat misleading. The function doesn't create any iterator, it appends the vectors.

**Review by FG-TUM — CHANGES_REQUESTED** [review]

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h (hunk 1 above)
Comment (dmartin13) [reply src/autopas/containers/verletClusterLists/VerletClusterLists.h]: done in [2cbb6b5](https://github.com/AutoPas/AutoPas/pull/753/commits/2cbb6b548bdbdf5c3fb851bb25a0170a172b0464)

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 2]
```diff
@@ -383,57 +404,135 @@ class VerletClusterLists : public ParticleContainerInterface<Particle>, public i
...
+   * @param additionalVectors
+   * @return true iff all passed vectors are empty
+   * @return false iff at least one buffer from the passed vectors is not empty
+   */
+  template <bool modifiable, bool regionIter>
+  bool additionalVectorsEmpty(
+      typename ContainerIterator<Particle, modifiable, regionIter>::ParticleVecType *additionalVectors) const {
+    bool addVectorsEmpty = true;
+    if (additionalVectors) {
+      for (auto &buffer : *additionalVectors) {
```
[the remark this answers is not in the document]
Comment (dmartin13) [reply src/autopas/containers/verletClusterLists/VerletClusterLists.h]: done in [2cbb6b5](https://github.com/AutoPas/AutoPas/pull/753/commits/2cbb6b548bdbdf5c3fb851bb25a0170a172b0464)

File: src/autopas/containers/verletClusterLists/VerletClusterLists.h [hunk 3]
```diff
@@ -383,57 +404,135 @@ class VerletClusterLists : public ParticleContainerInterface<Particle>, public i
...
+   * @tparam modifiable
+   * @tparam regionIter
+   * @param additionalVectors
+   * @return true iff all passed vectors are empty
+   * @return false iff at least one buffer from the passed vectors is not empty
+   */
+  template <bool modifiable, bool regionIter>
+  bool additionalVectorsEmpty(
+      typename ContainerIterator<Particle, modifiable, regionIter>::ParticleVecType *additionalVectors) const {
+    bool addVectorsEmpty = true;
```
[the remark this answers is not in the document]
Comment (dmartin13) [reply src/autopas/containers/verletClusterLists/VerletClusterLists.h]: done in [2cbb6b5](https://github.com/AutoPas/AutoPas/pull/753/commits/2cbb6b548bdbdf5c3fb851bb25a0170a172b0464)

**Comment by dmartin13** [comment]
I tried the solution where we add the particles to the towers right away instead of adding them to `_particlesToAdd` first. However, besides performance problems, this then also causes problems in `getRegionIterator` (which is called before the rebuild), since here it is already expected that the particles are sorted correctly into the towers. So I think the easiest solution at the moment is to just remove the unnecessary buffer allocations in the non-rebuild iterations, since `_particlesToAdd` has no particles here.

What else I could do would be in the functions `updateHaloParticle`, `forEach`, `reduce`, `forEachInRegion`, `reduceInRegion` and `deleteAllParticles`, to just remove the iterating over `_particlesToAdd`, since we never call those functions if there are particles in `_particlesToAdd`. But this would just reduce the code a bit and possibly improve readability. However, it would kind of break the logic. Just because we don't currently call these functions in the LogicHandler doesn't mean we can't call them at some point. What do you think?

**Comment by FG-TUM** [comment]
> I tried the solution where we add the particles to the towers right away instead of adding them to `_particlesToAdd` first. However, besides performance problems...

Since performance is one of our main concerns here this is a showstopper for this approach.

> So I think the easiest solution at the moment is to just remove the unnecessary buffer allocations in the non-rebuild iterations ...

I agree we should do it this way.
 
> What else I could do would be in the functions [...] However, it would kind of break the logic. Just because we don't currently call these functions in the LogicHandler doesn't mean we can't call them at some point. What do you think?

Again I agree. This seems to decrease the versatility of these functions and makes it easy to introduce bugs in the future, so I'd advise against it.

**Review by FG-TUM — APPROVED** [review]

**Pull request merged**